server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java

```diff
@@ -101,7 +106,7 @@ public void getSegmentFiles(
                 return;
             }
             logger.trace("Downloading segments files from remote store {}", filesToFetch);
-
+            cancellableThreads.checkForCancel();
```

Comment:
This check happens before any files are fetched from the store.  What happens if cancel comes after its started?  If there is no way to stop the ongoing download directly on the store I think we could instead use a sync cancel here similar to [recovery](https://github.com/opensearch-project/OpenSearch/blob/de2b6b7a6b7cafea204f1d9b3ff0830780bafc08/server/src/main/java/org/opensearch/indices/recovery/RecoverySourceHandler.java#L828) by wrapping the source calls in cancellableThreads.execute inside the target itself?  That way we don't need any of these signature changes to pass down the cancellablethreads.

Reply:
> This check happens before any files are fetched from the store. What happens if cancel comes after its started?

The call for `checkForCancel()` is also present inside loop where we download files. Thus, when execution pass this statement, it would halt download of subsequent files.

> If there is no way to stop the ongoing download directly on the store I think we could instead use a sync cancel here similar to [recovery](https://github.com/opensearch-project/OpenSearch/blob/de2b6b7a6b7cafea204f1d9b3ff0830780bafc08/server/src/main/java/org/opensearch/indices/recovery/RecoverySourceHandler.java#L828) by wrapping the source calls in cancellableThreads.execute inside the target itself?

This is a good idea around simplification. Though, above change would also apply to `PrimaryShardReplicationSource` source which perfroms async calls (via RetryableAction). Even though, it would be a NO-OP but would be mis-leading. With existing `checkForCancel` inside target, it might be confusing. I think we can continue to pass cancellableThreads to RemoteStoreReplicationSource and have execute wrapped inside cancellableThreads.execute(). WDYT ?

Reply:
I would prefer to not have to init and pass CancellableThreads all the way down.  I think if its *ok* for Node-node (we can validate with the suite test) so i would err on the side of simplicity.  We may need to swap the target to first invoke cancel on the source and then cancel the threads.

Reply:
Sure @mch2 I can go ahead with this change. I will add comments to avoid any confusion.